### DISCUSSION TITLE
Interface Design

### DISCUSSION THREAD

**Discussion by marcqualie** [tokenable/tokenable-ruby#2] [Ideas]
High level goal for how we want this gem to be used inside a rails application.

## Installation

```ruby
# Add to Gemfile
gem 'tokenable'

# Add route mount
mount Tokenable::AuthEndpoint => '/api/auth'

# Include the controller helper (could this be auto injected to ActionController::Base?)
class ApplicationController < ActionController::Base
  include Tokenable::Authable
end

# Include the user module for auto authentication
class User < ApplicationRecord
  include Tokenable::User
end
```

## Usage

A user would post their credentials to the `/api/auth` endpoint, which would return a JWT token or a 401 response if not authorised.

### Controllers

A controller has access to `user_signed_in? (boolean)` and `current_user (User|nil)`. These are also exposed to views.

```ruby
class WelcomeController < ApplicationController
  before_action :require_tokenable_user!
end
```

**Comment by scottrobertson** [comment]
This all looks good. 

And then at some point down the road, we can have a config file for `tokenable.user_class = 'SomeOtherUser'` for example.
